packages/engine/src/worker/runner/javascript/mod.rs

```diff
@@ -235,128 +163,168 @@ fn read_file(path: &str) -> Result<String> {
+                false,
+                false,
+                true,
+            )),
+        );
+        let mut try_catch_scope = v8::TryCatch::new(scope);
+        let module =
+            v8::script_compiler::compile_module(&mut try_catch_scope, source).ok_or_else(|| {
+                let exception = read_exception(&mut try_catch_scope).unwrap();
+
+                Error::Eval(path.to_string(), format!("Compile error: {exception}"))
+            })?;
+
+        module
+            .instantiate_module(&mut try_catch_scope, module_resolve_callback)
+            .ok_or_else(|| {
+                let exception = read_exception(&mut try_catch_scope).unwrap();
 
-    let global_context = context.global(scope);
-    let imported = call_js_function(scope, f, global_context.into(), args)
-        .map_err(|err| Error::FileImport(path.into(), format!("Could not call function: {err}")))?;
+                Error::PackageImport(
+                    path.to_string(),
+                    format!("Could not instantiate code for package: {exception}"),
+                )
+            })?;
+
+        if module.get_status() != v8::ModuleStatus::Instantiated {
+            return Err(Error::PackageImport(
+                path.to_string(),
+                "Could not instantiate code for package".to_string(),
+            ));
+        }
+
+        module.evaluate(&mut try_catch_scope).ok_or_else(|| {
+            let exception = read_exception(&mut try_catch_scope).unwrap();
+
+            Error::PackageImport(
+                path.to_string(),
+                format!("Could not evaluate code for package: {exception}"),
+            )
+        })?;
+
+        // `v8::Module::evaluate` can return `Some` even though the evaluation didn't
+        // succeed
+        if module.get_status() != v8::ModuleStatus::Evaluated {
+            let exception = module.get_exception();
+            let exception_string = exception
+                .to_string(&mut try_catch_scope)
+                .unwrap()
+                .to_rust_string_lossy(&mut try_catch_scope);
+
+            return Err(Error::PackageImport(
+                path.to_string(),
+                format!("Could not evaluate code for package: {exception_string}"),
+            ));
+        }
+
+        module_map.borrow_mut().path_to_module.insert(
+            path.to_string(),
+            v8::Global::new(&mut try_catch_scope, module),
+        );
+
+        Ok(module)
+    }
+}
 
-    Ok(imported)
+/// Callback called for each `import ...` in JS files. It will read the file, compile the source
+/// code and evaluate it. Modules are only compiled and evaluated once.
+// Simple example without any caching: https://gist.github.com/surusek/4c05e4dcac6b82d18a1a28e6742fc23e
+// More elaborate example with caching and multiple types of imports: https://github.com/denoland/deno/blob/f7e7f548499eff8d2df0872d1340ddcdfa028c45/core/bindings.rs#L1344
+pub fn module_resolve_callback<'s>(
```

Comment:
Shouldn't be public as we rely on this being a callback only. Alternative mark it as unsafe and delegate the safety requirements to the caller.

Reply:
Removed https://github.com/hashintel/hash/pull/493/commits/9b6c6b44209d59a3477b5b0c53d5b7f3d0739baa